### PR TITLE
feat: Load only required environments

### PR DESCRIPTION
**Pull request by Laubi**
Summary:
1) Extend LoadManifest to let consumers specify what environments and environment groups to load.
2) Extend the console command flag `--group` to allow multiple values (Deploy/Delete)

User-visible changes:
* Console commands:
  * Deploy/Delete: Change '--group' to allow multiple groups
* Manifest:
  * Users may (again) specify environment variables that can not be resolved if the environment is not required to load. 

Library changes:
* LoadManifest now accepts Environments and Groups that should be loaded. If they can't be found, LoadManifest errors. Environments not included in this group are not parsed and might be invalid (wrong configuration, env vars not set, etc)

### CODE REVIEW

File: pkg/manifest/manifest_loader.go [hunk 1]
```diff
@@ -307,20 +325,46 @@ func toEnvironments(context *ManifestLoaderContext, groups []group) (map[string]
...
+			}
+
+			environments[parsedEnv.Name] = parsedEnv
+		}
+	}
+
+	// validate that all required groups & environments are included
+	for _, g := range context.Groups {
+		if !groupNames[g] {
+			errors = append(errors, newManifestLoaderError(context.ManifestPath, fmt.Sprintf("required to load group %q, but could not found", g)))
```
Comment (UnseenWizzard) [inline pkg/manifest/manifest_loader.go:361]:
```suggestion
			errors = append(errors, newManifestLoaderError(context.ManifestPath, fmt.Sprintf("requested group %q not found", g)))
```

File: pkg/manifest/manifest_loader.go [hunk 2]
```diff
@@ -307,20 +325,46 @@ func toEnvironments(context *ManifestLoaderContext, groups []group) (map[string]
...
+		}
+	}
+
+	for _, e := range context.Environments {
+		if !envNames[e] {
+			errors = append(errors, newManifestLoaderError(context.ManifestPath, fmt.Sprintf("required to load environment %q, but could not found", e)))
```
Comment (UnseenWizzard) [inline pkg/manifest/manifest_loader.go:367]:
```suggestion
			errors = append(errors, newManifestLoaderError(context.ManifestPath, fmt.Sprintf("requested environment %q not found", e)))
```

alternative language fix: "but could not find it" - but I think these messages can be shorter and still say what's wrong clearly

File: cmd/monaco/deploy/command.go [hunk 3]
```diff
@@ -47,12 +47,12 @@ func GetDeployCommand(fs afero.Fs) (deployCmd *cobra.Command) {
...
 	}
 
-	deployCmd.Flags().StringSliceVarP(&environment, "environment", "e", make([]string, 0), "Specify one (or multiple) environments to deploy to. To set multiple environments either repeat this flag, or seperate them using a comma (,). This flag is mutually exclusive with '--group'.")
-	deployCmd.Flags().StringVarP(&group, "group", "g", "", "Specify the environmentGroup that should be used for deployment. If this flag is specified, all environments within this group will be used for deployment. This flag is mutually exclusive with '--environment'")
+	deployCmd.Flags().StringSliceVarP(&environment, "environment", "e", []string{}, "Specify one (or multiple) environment(s) to deploy to. To set multiple environments either repeat this flag, or separate them using a comma (,). This flag is mutually exclusive with '--group'.")
+	deployCmd.Flags().StringSliceVarP(&groups, "group", "g", []string{}, "Specify one (or multiple) environmentGroup(s) to deploy to. To set multiple groups either repeat this flag, or seprate them using a comma (,). If this flag is specified, all environments within this group will be used for deployment. This flag is mutually exclusive with '--environment'")
```
Comment (UnseenWizzard) [inline cmd/monaco/deploy/command.go:55]: did we already wrongly describe this as repeatable in the documentation, or will we have to update it there as well?

File: cmd/monaco/deploy/deploy.go [hunk 4]
```diff
@@ -34,40 +34,35 @@ import (
...
 )
 
-func deployConfigs(fs afero.Fs, manifestPath string, environmentGroup string, specificEnvironments []string, specificProjects []string, continueOnErr bool, dryRun bool) error {
+func deployConfigs(fs afero.Fs, manifestPath string, environmentGroups []string, specificEnvironments []string, specificProjects []string, continueOnErr bool, dryRun bool) error {
 	absManifestPath, err := absPath(manifestPath)
 	if err != nil {
 		return fmt.Errorf("error while finding absolute path for `%s`: %w", manifestPath, err)
 	}
-	loadedManifest, err := loadManifest(fs, absManifestPath)
+	loadedManifest, err := loadManifest(fs, absManifestPath, environmentGroups, specificEnvironments)
```
Comment (UnseenWizzard) [inline cmd/monaco/deploy/deploy.go:42]: nitpick: This commit does a lot more than allowing multiple groups to be defined. More notably it replaces the external environment filtering with the manifest loader filtering. 

If you want nice atomic commits I'd split into 1) use the manifestLoader filtering, 2) allow multiple groups.
If you don't care for that, update the message so it's clear from the commit message what changed

File: cmd/monaco/delete/command.go [hunk 5]
```diff
@@ -52,13 +52,24 @@ func GetDeleteCommand(fs afero.Fs) (deleteCmd *cobra.Command) {
...
+			"To set multiple groups either repeat this flag, or separate them using a comma (,). "+
+			"If this flag is specified, all environments within these groups will be used for deployment. "+
+			"This flag is mutually exclusive with '--environment'. "+
+			"If this flag is specified, configuration will be deleted from all environments within the specified groups. "+
+			"If neither --groups nor --environment is present, all environments will be used for deletion")
+	deleteCmd.Flags().StringSliceVarP(&environments, "environment", "e", []string{},
+		"Specify one (or multiple) environments(s) that should be used for deletion. "+
+			"To set multiple environments either repeat this flag, or separate them using a comma (,). "+
+			"This flag is mutually exclusive with '--group'. "+
+			"If this flag is specified, configuration will be deleted from all environments within the specified groups. "+
```
Comment (UnseenWizzard) [inline cmd/monaco/delete/command.go:71]:
```suggestion
			"If this flag is specified, configuration will be deleted from all specified environments. "+
```

File: cmd/monaco/delete/command.go [hunk 6]
```diff
@@ -52,13 +52,24 @@ func GetDeleteCommand(fs afero.Fs) (deleteCmd *cobra.Command) {
...
 
-			return Delete(fs, manifestName, deleteFile, environments, group)
+			return Delete(fs, manifestName, deleteFile, environments, groups)
 		},
 		ValidArgsFunction: completion.DeleteCompletion,
 	}
 
-	deleteCmd.Flags().StringVarP(&group, "group", "g", "", "Specify the environmentGroup that should be used for deletion. This flag is mutually exclusive with '--environment'. If this flag is specified, configuration will be deleted from all environments within the specified group.")
-	deleteCmd.Flags().StringSliceVarP(&environments, "environment", "e", make([]string, 0), "Deletes configuration only for specified environments. This flag is mutually exclusive with '--group' If not set, delete will be executed on all environments defined in manifest.")
+	deleteCmd.Flags().StringSliceVarP(&groups, "group", "g", []string{},
```
Comment (UnseenWizzard) [inline cmd/monaco/delete/command.go:60]: like deploy, do we need to update the documentation?

File: pkg/manifest/manifest.go [hunk 7]
```diff
@@ -141,16 +141,3 @@ func (e Environments) FilterByNames(names []string) (Environments, error) {
 
 	return result, nil
 }
-
-// FilterByGroup returns all environments whose group-name matches the given name.
-func (e Environments) FilterByGroup(groupName string) Environments {
```
Comment (UnseenWizzard) [inline pkg/manifest/manifest.go:146]: commit nitpick: why is it removed? (In other commits you state 'unused function' in the message)

File: pkg/manifest/manifest_test.go [hunk 8]
```diff
@@ -1,113 +0,0 @@
...
-
-import (
-	"github.com/google/go-cmp/cmp/cmpopts"
-	"gotest.tools/assert"
-	"testing"
-)
-
-var sortStrings = cmpopts.SortSlices(func(a, b string) bool { return a < b })
-
-func TestManifestFilterEnvironmentsByNamesWithEmptyNames(t *testing.T) {
```
Comment (UnseenWizzard) [inline pkg/manifest/manifest_test.go:26]: you've covered all of these cases with the ones you added for the filtering on load right?

File: cmd/monaco/deploy/command.go (hunk 3 above)
Comment (Laubi) [reply cmd/monaco/deploy/command.go:55]: Currently, I would argue that the docs are wrong: `Apply your configurations to specific environment groups within your deployment manifest file.` (groups!)
With this change, the behavior is as the docs state it. Not sure on what level we want to keep the docs and Monaco-command-descriptions aligned. 

https://www.dynatrace.com/support/help/manage/configuration-as-code/reference/commands

File: cmd/monaco/delete/command.go [hunk 9]
```diff
@@ -52,13 +52,24 @@ func GetDeleteCommand(fs afero.Fs) (deleteCmd *cobra.Command) {
...
+		"Specify one (or multiple) environmentGroup(s) that should be used for deletion. "+
+			"To set multiple groups either repeat this flag, or separate them using a comma (,). "+
+			"If this flag is specified, all environments within these groups will be used for deployment. "+
```
[the remark this answers is not in the document]
Comment (Laubi) [reply cmd/monaco/delete/command.go:63]: 🤦

File: cmd/monaco/deploy/command.go (hunk 3 above)
Comment (UnseenWizzard) [reply cmd/monaco/deploy/command.go:55]: I think I probably documented that wrongly thinking that groups are like environments and copying the description. 

As for alignment, I'd say "generally correct" doesn't need to match wording but shouldn't be wrong and shouldn't be missing new commands/flags

File: pkg/manifest/manifest_test.go (hunk 8 above)
Comment (Laubi) [reply pkg/manifest/manifest_test.go:26]: Yes

File: pkg/manifest/manifest.go (hunk 7 above)
Comment (Laubi) [reply pkg/manifest/manifest.go:146]: reworded

File: cmd/monaco/deploy/deploy.go (hunk 4 above)
Comment (Laubi) [reply cmd/monaco/deploy/deploy.go:43]: rebased both deploy and delete commits

File: cmd/monaco/deploy/command.go (hunk 3 above)
Comment (Laubi) [reply cmd/monaco/deploy/command.go]: 👍

**Review by UnseenWizzard — APPROVED** [review]

**Pull request merged**